src/transformers/quantizers/quantizer_torchao.py

```diff
@@ -0,0 +1,157 @@
+        param_name: str,
+        state_dict: Dict[str, Any],
+        **kwargs,
+    ) -> bool:
+        module, tensor_name = get_module_from_name(model, param_name)
+
+        return isinstance(module, torch.nn.Linear) and (tensor_name == "weight")
+
+    def create_quantized_param(
+        self,
+        model: "PreTrainedModel",
+        param_value: "torch.Tensor",
+        param_name: str,
+        target_device: "torch.device",
+        state_dict: Dict[str, Any],
+        unexpected_keys: List[str],
+    ):
+        """
+        Each nn.Linear layer is processsed here.
+        We first check if the corresponding module state_dict contains already torchao quantized parameters.
+        If not, we create a temp linear layer with the module state_dict params and use it for quantization
+        """
+        module, tensor_name = get_module_from_name(model, param_name)
+
+        layer_name = param_name.replace(".weight", "").replace(".bias", "")
+        parent_module = find_parent(model, layer_name)
+        node = layer_name.split(".")[-1]
+
+        # Step 0: set module state_dict
+        module_state_dict = {key.split(".")[-1]: state_dict[key] for key in state_dict if layer_name in key}
+
+        # Step 1: populate module with weight/bias from module state dict
+        for key in module_state_dict:
+            setattr(module, key, torch.nn.Parameter(module_state_dict[key]))
+
+        # Step 2: Replace module with either HQQLinear or move it to device. We do this via setattr on the parent as doing on it on the module
+        # directly doesn't work.
+
+        _STR_TO_METHOD = {
+            "int4_weight_only": int4_weight_only,
+            "int8_weight_only": int8_weight_only,
+            "int8_dynamic_activation_int8_weight": int8_dynamic_activation_int8_weight,
+        }
+
+
+        module = module.to(dtype=self.torch_dtype, device=target_device)
+        setattr(parent_module, node, module)
+
+        if self.quantization_config is not None:
+            assert self.quantization_config.quant_type in _STR_TO_METHOD.keys(), f"Requested quantization type: {self.quantization_config.quant_type} is not supported yet, please add support in TorchAOHfQuantizer."
+            quantize_(module, _STR_TO_METHOD[self.quantization_config.quant_type](**self.quantization_config.kwargs))
+            setattr(parent_module, node, module)
+
+        torch.cuda.empty_cache()
+
+    def _process_model_before_weight_loading(self, model: "PreTrainedModel", **kwargs):
+        """No process required for torchao quantized model
+        """
+        return
+
+    def _process_model_after_weight_loading(self, model):
+        """No process required for torchao quantized model
+        """
+        return
+
+    @property
+    def is_serializable(self):
+        return False
+
+    @property
+    def is_trainable(self):
+        # torchao does not have official support for QAT (Quantization Aware Training or PEFT yet.)
```

Comment:
We do have support for nf4 and QAT though, is the intent here to say the hf ao backend does not support this yet?

Reply:
QAT is still in protootype I think, oh I forgot about nf4, I haven't thought of how we expose nf4 yet, will check with Driss

Reply:
I think we haven't tested nf4 dtype with quantize_ API yet, maybe we can create an issue to make nf4 tensor work with `quantize_` API

Reply:
checked with Driss, we probably need to do a separate integration for nf4/QLoRA like what we did for torchtune, need to implement a (Q)LoRALinear class like this https://github.com/pytorch/torchtune/blob/288ff4435b0cf17325b5c3b112f6859a6cdf0ea2/torchtune/modules/peft/lora.py#L18 in huggingface